compiler/rustc_codegen_ssa/src/debuginfo/type_names.rs

```diff
@@ -273,55 +405,143 @@ pub fn push_debuginfo_type_name<'tcx>(
 
             output.push_str("enum$<");
             push_item_name(tcx, def.did, true, output);
-            push_type_params(tcx, substs, output, visited);
+            push_generic_params_internal(tcx, substs, false, output, visited);
 
             let dataful_variant_name = def.variants[*dataful_variant].ident.as_str();
 
             output.push_str(&format!(", {}, {}, {}>", min, max, dataful_variant_name));
         } else {
             output.push_str("enum$<");
             push_item_name(tcx, def.did, true, output);
-            push_type_params(tcx, substs, output, visited);
-            output.push('>');
+            push_generic_params_internal(tcx, substs, false, output, visited);
+            push_close_angle_bracket(tcx, output);
         }
     }
+}
+
+pub fn push_item_name(tcx: TyCtxt<'tcx>, def_id: DefId, qualified: bool, output: &mut String) {
+    let def_key = tcx.def_key(def_id);
+    if qualified {
+        if let Some(parent) = def_key.parent {
+            push_item_name(tcx, DefId { krate: def_id.krate, index: parent }, true, output);
+            output.push_str("::");
+        }
+    }
+
+    push_unqualified_item_name(tcx, def_id, def_key.disambiguated_data, output);
+}
+
+fn push_unqualified_item_name(
+    tcx: TyCtxt<'tcx>,
+    def_id: DefId,
+    disambiguated_data: DisambiguatedDefPathData,
+    output: &mut String,
+) {
+    let cpp_like_names = tcx.sess.target.is_like_msvc;
 
-    fn push_item_name(tcx: TyCtxt<'tcx>, def_id: DefId, qualified: bool, output: &mut String) {
-        if qualified {
+    match disambiguated_data.data {
+        DefPathData::CrateRoot => {
             output.push_str(&tcx.crate_name(def_id.krate).as_str());
-            for path_element in tcx.def_path(def_id).data {
-                write!(output, "::{}", path_element.data).unwrap();
+        }
+        DefPathData::Impl => {
+            let self_ty = tcx.type_of(def_id);
+            output.push_str(if cpp_like_names { "impl$<" } else { "<impl " });
+            if let Some(impl_trait_ref) = tcx.impl_trait_ref(def_id) {
+                push_item_name(tcx, impl_trait_ref.def_id, true, output);
+                output.push_str(if cpp_like_names { ", " } else { " for " });
             }
-        } else {
-            output.push_str(&tcx.item_name(def_id).as_str());
+            push_debuginfo_type_name(tcx, self_ty, false, output, &mut FxHashSet::default());
```

Comment:
I think we want the qualified version of the self type here, right?

Reply:
We already have the qualifier, so I don't think that we need to add it again here.

Current format:
```
a!function_names::Mod1::impl$<function_names::Mod1::TestTrait2, TestStruct2>::trait_function
```

If we qualified the self type:
```
a!function_names::Mod1::impl$<function_names::Mod1::TestTrait2, function_names::Mod1::TestStruct2>::trait_function
```

Unless you wanted to add in some special casing so that the qualifier is moved to the self type:
```
a!impl$<function_names::Mod1::TestTrait2, function_names::Mod1::TestStruct2>::trait_function
```

Reply:
The path to the impl is actually independent of the path to the self type, as in this [example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=2d8bddc90fea0703e8e5d829302711ae):

```rust

pub mod foo {
    pub struct X;
}

mod bar {
    impl super::foo::X {
        pub fn bar() {}
    } 
}
```

So the version that does not print the path to the impl is actually the correct one. The symbol mangling RFC has [a section](https://rust-lang.github.io/rfcs/2603-rust-symbol-name-mangling-v0.html#methods) describing the situation in a bit more detail.

Reply:
Huh, interesting - I'll switch to qualifying the self type instead of the impl